### ISSUE TITLE
Integration tests for proper boot verification

### ISSUE THREAD

**Issue by Rudd-O** [zfsonlinux/zfs#4555]
You know what we need?

We need something that runs https://github.com/Rudd-O/zfs-fedora-installer for each Fedora release beginning with Fedora 17 or something, and validates a particular combo of ZFS and SPL repos, to see if they boot.  The program has a simulated-with-QEMU boot process to create an image that can be later tested again with QEMU.

We need this program to be somehow incorporated to the buildbot runs, and we need to add to it so it also tests that the built image boots properly.

We need to run this periodically as well, because some distros move relatively fast, and sometimes their changes break booting from ZFS / Dracut through no fault of our own.

I can do the necessary work on the zfs-fedora-installer side.  I am even willing to figure out if we can extend it to other distros.

**Comment by dilshath** [comment]
I am using centos 7 in physical machine. While creating and saving new partition followed by using "fdisk" command for save and partprobe in Centos 7 (It is installed in KVM), I got this error (at least 3 or 4 times). please guide me how to fix this problem.

Warning: /dev/centos/root does not exist
Warning: /dev/centos/swap does not exist
Warning: /dev/mapper/centos-root does not exist

Starting Dracut Emergency Shell...
Warning: /dev/centos/root does not exist
Warning: /dev/centos/swap does not exist
Warning: /dev/mapper/centos-root does not exist

Generating "/run/initramfs/rdsosreport.txt"
Entering emergency mode, Exit the shell to continue.

dracut:/#

**Comment by Rudd-O** [comment]
Your bug report refers to an ext4 LVM boot.  It has nothing to do with ZFS.

Please do not post unrelated issues to bugs here.

**Comment by Rudd-O** [comment]
Here is what I think the plan needs to be, from an inner loop perspective.  To test a pair of spl / zfs commits:

* Start Amazon EC2 VM.  Use user-data to add SSH key.
* Clone SPL and ZFS repos, check out commits in question.
* Build RPM-dkms set for both.
* SSH into the box to run zfs-fedora-installer, but pointing it to the built RPM set in question.
* Wait till the process is done.
* If the process fails, report error.
* If the process succeeds, attempt to boot the resulting image within QEMU (not the KVM variant of QEMU, obviously), scripting it with expect, so that the command line launches a /sbin/init that is really just a "echo OK, now poweroff".
* If the boot completes in a certain amount of time and succeeds in producing the "success!" output, then declare the commits as good.
* If the boot fails, then declare the commits as bad.
* Tear down the VM unconditionally.

Of course, the outer (control) loop for this process needs to be figured out as well.  But that we can discuss later.  Right now, this inner loop is necessary even for manual testing.

**Comment by behlendorf** [comment]
If this can **all** be done under a running ec2 instance (without rebooting) we should be add it to the existing buildbot infrastructure.  The buildbot operates today as the outer loop by provisioning a clean VM, booting it, running a set of scripts, reporting the results, and then destroying the VM.  The major thing the buildbot can't handle is rebooting the VM where a buildslave is running.  That will be treated as a failure.

So if you can write a script, or set of scripts, which perform your proposed inner loop we can add it.  It's worth taking a look at the [existing test scripts](https://github.com/zfsonlinux/zfs-buildbot/tree/master/scripts) used by the buildbot.  Each of these simple scripts is called as a separate build step by what's known as a buildbot factory.  Additional factories can be added to the [master config](https://github.com/zfsonlinux/zfs-buildbot/blob/master/master/master.cfg) so for testing dracut we can define a new one which calls the required scripts.

**Comment by Rudd-O** [comment]
Yes, it can be done.  How does the buildbot feed data (parameters / scripts) to the clean VM?  I need to know in order to write a script that, at the very least, will attempt to build a bootable image (we can later add support for testing said image).

**Comment by Rudd-O** [comment]
My question goes more like "so, this buildbot, how does the testing system give it the exact commits of SPL and ZFS to test?" and "how do I get the buildbot to download the zfs-fedora-installer repo as well?"

Basically that.

**Comment by Rudd-O** [comment]
I see how the commits are taken: https://github.com/zfsonlinux/zfs-buildbot/blob/fa1a8c14f5ee3e805545474dff1523d62165bd36/master/master.cfg#L698

Eeeenteresting.  Alright.  I will see how the other scripts do their work.

**Comment by Rudd-O** [comment]
Having read the scripts, it's not clear to me how to write a new script that will check out zfs-fedora-installer, cd into its directory, and then, using the DKMS RPMs that were built before by the build system, attempt to build an image.  In fact, none of the scripts build SPL and ZFS at once, nor do they seem to build DKMS packages either (make pkg is what i see).

**Comment by Rudd-O** [comment]
I believe I have found the thing: https://github.com/zfsonlinux/zfs-buildbot/blob/fa1a8c14f5ee3e805545474dff1523d62165bd36/master/master.cfg#L151

**Comment by Rudd-O** [comment]
https://github.com/zfsonlinux/zfs-buildbot/pull/25

Please let me know if this is not adequate or if it is a huge fuckup.

Later on, we can figure out how to make it so that we can verify that the image booted.  That may require some qemu-system-x86 and expect scripting.  Painful, but it should probably be doable.  Maybe we declare success if the thing gets to the boot prompt, detecting that circumstance with expect.

**Comment by behlendorf** [comment]
Looks like an excellent start we'll just need to iterate a little bit to get everything working smoothly.

**Comment by Rudd-O** [comment]
Happy to do so.  I will heed your requests in the following days.

Do you come to Europe often?  It'd be nice to hash out a few details of future features and fixes over a few beers.

**Comment by behlendorf** [comment]
@Rudd-O not as often as I'd like!  I'll see about giving the current version of the script some manual testing to make sure it works as expected.  What distributions should I expect it to work for, just the latest Fedora?

**Comment by Rudd-O** [comment]
The script, as is, invoking zfs-fedora-installer, is only going to work for Fedoras that are 17 and up (or at least that's where I started with it, and it worked back then, and there were no substantial changes that would break it with Fedora 17).

Unfortunately, I use some "hard coded" repo files, but we can always provide extra repo files for other distributions that aren't Fedora (CentOS, Red Hat), and those repo files could be selected with a command line argument... it *should* work.  The dichotomy is always the host system versus the chroot system, where the host system can install packages with yum/dnf, but the guest system may have to use dnf/yum to manipulate the packages.  I haven't cross tested, say, Red Hat 6 as a guest with Fedora 23 as a host, or vice versa, but I have tested yum on the host and DNF on the chroot, and that worked fine.  These are minor points that we can hash out by trying them — and I will try my best to get these to work out here in the local Xen setup (equivalent to Amazon EC2).

Let's see about those beers.
